### ISSUE TITLE
Add better support for algorithm versions

### ISSUE THREAD

**Issue by depryf** [imsweb/algorithms#132]
Right now the algorithm mechanism only allows on algorithm per version. 

This is starting to cause some issues.

I would like to investigate a way to support multiple versions for algorithms.

At first, I thought I would just add a version parameter to the ones needed it. It's a simple change, and it wouldn't affect much. But there is a use-case that it doesn't cover: sometimes different versions of an algorithm needs to write to a difference fields (for example alg-v1 needs to set field1 and alg-v2 needs to set fields2 so that they can be run concurrently and the two fields can be compared. 

It's also possible that a new version of an algorithm requires new input fields. 

If supporting those use-cases is a must, then the different versions need to be implement as their own algorithms. 

That means if an application wants to only show one of them, it will need to filter them out (I might provide a way to do that via the library).

I need to spend a bit more time to fully understand the side effects of having multiple algorithms for computing the same output variables.

**Comment by depryf** [comment]
@howew I am notifying you on this issue because I know it will affect your software.

**Comment by depryf** [comment]
The more I think about it, the more I dislike the idea of duplicating the same algorithm so it supports multiple versions.

From a logical point of view, I don't think it makes sense to allow the same algorithms to write to two different fields based on the version. I understand the use case is to be able to compare the two variables (have them together in the same data set), but I think this is a rare use-case and there are ways to still achieve that with a single variable (compute the two variables in two different files and then compare the fields). 

I think having the version as a parameter makes more sense and will be more user-friendly (and application-friendly). I do want to support a default value for that param though, which will be the "latest version available".

**Comment by howew** [comment]
Thinking outside the box... I don't know if this is doable or not, but could we have some sort of a plugin system where users of File*Pro or what have you could add versions of the algorithms library jar to the class path?

**Comment by depryf** [comment]
Dynamically registering algorithms would solve that use-case (in the sense that a user could register whatever algorithm, with whatever version and set whatever output field) , but that is a pretty complex feature to put in place.

I am not worried about algorithms that need to compute different versions into different fields; I think that's a rare use-case.

Here is my current thought: I am going to add a "version" parameter to the algorithms that need to support more than one version. If the version is not provided, it will use the "default" one (which is usually the latest one). 

For an application like NAACCR*Prep, there will be no changes required (although if you dynamically show all the parameters, you might need to ignore the "version" one).  For an application like File*Pro, it will display the version like any other parameter, and allow the user to select one of them.  I think this path is really not intrusive to what we currently have.

**Comment by depryf** [comment]
So I started to implement this with a new "version" parameter, but now I am not so sure this is the best approach.

Most of the algorithms that need multiple versions (like ICCC or AYA) work the same way: they take site/hist/beh and compute a code. The problem is that the computed code has a different meaning in each of those versions.   And so when we add the value to a data file, and the value says "01",  one has to know the version to know that "01" means. 

An even stronger example is the new AYA version where the recode went from 2 digits to 3.

Because of those issues, I am starting to think the correct approach is to duplicate the algorithms...

**Comment by depryf** [comment]
Moving forward, different versions will create different algorithms. 

For this issue, I removed support for the "info" attribute since there was a lot of confusion about what that really was (sometimes it was left empty, sometimes it was set to the name, and sometimes it provided a long English description). To get an algorithm version, just concatenate the name and version, with a space in between.

I also tweaked a few algorithm names to better represent their "ownership".

That's all the cleanup I want to do for this issue.
